File: Source Files/UNITY C#/Assets/Scripts/NEAT/Agent.cs
using UnityEngine;

namespace Assets.Scripts.NEAT
{
    public class Agent
    {
        private Color color;
        private float maxSpeed;
        private Vector3 pos, vel, acc;

        public bool Dead { get; set; }
        public bool IsBest { get; set; }
        public GameObject Sphere { get; set; }
        public Brain Brain { get; private set; }
        public float Fitness { get; private set; }
        public bool ReachedGoal { get; private set; }
        public Color Color { get => color; set => color = value; }

        public Agent(int directionArraySize, float mutationRate, float speed, float maxSpeed)
        {
            Dead = false;
            IsBest = false;
            ReachedGoal = false;

            Fitness = 0;
            this.maxSpeed = maxSpeed;

            Brain = new Brain(directionArraySize, mutationRate, speed);
            Sphere = Object.Instantiate(GameObject.FindWithTag("Start"));

            vel = Vector3.zero;
            acc = Vector3.zero;
            pos = Sphere.transform.position;
        }

        private void Move()
        {
            if (Brain.Directions.Length > Brain.Step)
            {
                acc = Brain.Directions[Brain.Step];
                Brain.IncStep();
            }
            else Dead = true;

            if (Mathf.Abs(vel.x + acc.x) < maxSpeed) vel.x += acc.x;
            if (Mathf.Abs(vel.y + acc.y) < maxSpeed) vel.y += acc.y;
            if (Mathf.Abs(vel.z + acc.z) < maxSpeed) vel.z += acc.z;
            pos += vel;

            Sphere.transform.position = pos;
        }

        public void Update()
        {
            if (!Dead && !ReachedGoal)
            {
                Move();
                if (Sphere.GetComponent<OnCollision>().touchedGoal)
                    ReachedGoal = true;
                else Dead |= Sphere.GetComponent<OnCollision>().touchedWall;
            }

            Sphere.GetComponent<MeshRenderer>().material.color = IsBest ? Color.green : color;
        }

        public void CalculateFitness()
        {
            if (ReachedGoal) Fitness = Brain.Directions.Length / (Brain.Step * Brain.Step);
            else Fitness = 1.0f / Mathf.Pow(GameObject.FindWithTag("Goal").GetComponent<Information>().GetDistance(Sphere), 2);
        }

        public Agent GetCopy() { return new Agent(Brain.Directions.Length, Brain.MutationRate, Brain.Speed, maxSpeed) { Brain = Brain.Clone() }; }
    };
}

File: Source Files/neat/src/Agent.cpp
//
// Created by Вадим Холошня on 17.03.2020.
//

#include "Agent.hpp"

Agent::Agent() : fitness(0.0f), dead(false), is_best(false), reached_goal(false),
                 pos(map->pos_agent), vel(0.0f, 0.0f), acc(0.0f, 0.0f)
{
    circle.setFillColor(sf::Color::Black);
    circle.setRadius(radius_agent);
    circle.setOrigin(radius_agent, radius_agent);
}

Agent Agent::get_copy()
{
    Agent copy;
    copy.brain = brain.clone();
    return copy;
}

void Agent::move()
{
    if (brain.directions.size() > brain.step)
    {
        acc = brain.directions[brain.step];
        brain.step++;
    } else dead = true;
    if (abs(vel.x + acc.x) < max_speed) vel.x += acc.x;
    if (abs(vel.y + acc.y) < max_speed) vel.y += acc.y;
    pos += vel;
}

void Agent::update()
{
    if (!dead && !reached_goal)
    {
        move();
        if (map->touched_additional_reward(pos)) fitness++;
        else if (map->touched_wall(pos)) dead = true;
        else if (map->touched_goal(pos)) reached_goal = true;
    }
}

void Agent::calculate_fitness()
{
    if (reached_goal) fitness = brain.directions.size() + 1.0f / static_cast<float>(brain.step * brain.step);
    else fitness = 1.0f / (map->dist(pos) * map->dist(pos));
}

void Agent::show(sf::RenderWindow &window)
{
    circle.setPosition(pos);
    window.draw(circle);
}
